#**EMAX:**

José Emiliano Díaz Gutiérrez | A01711141

Ximena Cantera Reséndiz       | A01277310

Ma. Fernanda Barba De Los Santos  | A01657868


#**Importar librerías y documentos**

In [2]:
import pandas as pd
import numpy as np
#import matplotlib.pyplot as plt
#import seaborn as sns


In [3]:
distribuidores = pd.read_excel("../data/ArchivosOriginales/Distribuidores_Todo.xlsx")
horas = pd.read_excel("../data/ArchivosOriginales/Horas 2024-2025.xlsx")
ruteo = pd.read_excel("../data/ArchivosOriginales/Indicadores de Ruteo.xlsx")
mantenimientos = pd.read_excel("../data/ArchivosOriginales/Mantenimientos 2024-2025.xlsx")
population = pd.read_excel("../data/ArchivosOriginales/PopulationView_2026.xlsx")
reporte = pd.read_excel("../data/ArchivosOriginales/Reporte_unidades_dia_anterior.xlsx", header=1)

In [4]:
distribuidores.columns

Index(['No.', 'Unidad', 'Distribuidor', 'Marca', 'Modelo', 'Estado', 'Ciudad',
       'Horas'],
      dtype='str')

In [5]:
horas.columns

Index(['Num', 'Distribuidor', 'IMEI', 'Alias', 'Marca', 'Modelo', 'Ciudad',
       'Estado', 'Latitud', 'Longitud', '2024-01', '2024-02', '2024-03',
       '2024-04', '2024-05', '2024-06', '2024-07', '2024-08', '2024-09',
       '2024-10', '2024-11', '2024-12'],
      dtype='str')

In [6]:
ruteo.columns

Index(['Division', 'Distancia del Dia', 'Distancia de Ruta',
       'Cumplimiento Salida (06:30:00)', '% Utilizacion', '% Visitas Plan',
       '% Visitas Clientes', 'Total Kilos'],
      dtype='str')

In [7]:
mantenimientos.columns

Index(['No', 'ALIAS', 'MARCA', 'MODELO', 'CONFIGURACION', 'NO SERIE',
       'DISTRIBUIDOR', 'SERVICIO', 'HRMTRO', 'ACTUAL', 'TIPO', 'FECHA', 'O.T.',
       'ESTATUS', 'Editar'],
      dtype='str')

In [8]:
population.columns

Index(['VIN (9 CHARACTERS)', 'VIN (17 CHARACTERS)', 'VIN', 'BRAND', 'TYPE',
       'PRODUCT', 'SERIES', 'MODEL', 'TECHNICAL TYPE', 'ICE CODE',
       'DEVICE SERIAL NUMBER', 'SEVERITY LEVEL', 'CONTEXT', 'DEALER',
       'DEALER NAME', 'DEALER CODE', 'COUNTRY OF COMPETENCE', 'LATITUDE',
       'LONGITUDE', 'LAST KNOWN POSITION (COUNTRY)', 'TELEMATICS BOX',
       'TELEMATICS BRAND', 'TELEMATICS TYPE', 'TELEMATICS MODEL', 'OWNER NAME',
       'ENGINE HOURS', 'LATEST COMMUNICATION DATE', 'BDE PROCESSED AT TIME',
       'LOWER PIN BREAK', 'UPPER PIN BREAK', 'MYFLEET DEALER',
       'MYFLEET DEALER NAME', 'EULA ACCEPTED', 'EULA DATE'],
      dtype='str')

In [9]:
reporte.columns

Index([                 'No',                'IMEI',               'Alias',
                     'Marca',              'Modelo',       'Configuracion',
                'Fecha Alta',           'No. Serie',        'Distribuidor',
                    'Ciudad',              'Estado',           'Horometro',
                  'Odometro',               'Total',            'Cerrados',
                   'C.Fuera',          'Pendientes',                    50,
                         300,                   600,                   900,
                        1200,                  1500,                  1800,
                        2100,                  2400,             'Cliente',
                  'Telefono',               'Email',             'Latitud',
                  'Longitud',                 'PEP',         'Combustible',
                 'Distancia',   'Vel. Maxima (kph)', 'Vel. Promedio (kph)',
        'T. Encendido (hrs)',    'T. Apagado (hrs)'],
      dtype='object')

#**Limpieza de datos**

**Nulos**

**Distribuidores**

In [10]:
distribuidores.isnull().sum()
#no hay nulos, no se borra nada

No.             0
Unidad          0
Distribuidor    0
Marca           0
Modelo          0
Estado          0
Ciudad          0
Horas           0
dtype: int64

In [11]:
distribuidores.duplicated().sum()
#no tiene duplicados

np.int64(0)

**Horas**

In [12]:
horas.isnull().sum()

Num                0
Distribuidor       0
IMEI               0
Alias              0
Marca             15
Modelo          6263
Ciudad            15
Estado          6263
Latitud           15
Longitud          15
2024-01            0
2024-02            0
2024-03            0
2024-04            0
2024-05            0
2024-06            0
2024-07            0
2024-08            0
2024-09            0
2024-10            0
2024-11            0
2024-12            0
dtype: int64

En esta base horas, las columnas más importantes son las de identificación y las columnas mensuales, porque sirven para observar el comportamiento mensual del equipo durante 2024 y 2025. Por lo que en esta base se busca analizar evolución mensual, actividad por equipo, comparativos entre años y concentración de actividad por región o distribuidor.

In [13]:
horas = horas.drop(columns=[
    'Num)',
    'Modelon',
    'Estado',
], errors='ignore')

En la base de datos de horas se eliminaron las columnas Num, Modelo y Estado, ya que no aportaban información relevante para el objetivo final del análisis.
La columna Num funciona únicamente como un consecutivo o índice interno, por lo que no representa una variable explicativa relacionada con el comportamiento operativo de las unidades. Por otro lado, las columnas Modelo y Estado presentaban una cantidad considerable de valores nulos, lo que podía afectar la calidad del análisis y generar ruido dentro del proceso de limpieza y modelado.

In [14]:
horas['Ciudad'] = horas['Ciudad'].fillna('Desconocido')
horas['Marca'] = horas['Marca'].fillna('Desconocido')
horas['Latitud'] = horas['Latitud'].fillna(0)
horas['Longitud'] = horas['Longitud'].fillna(0)

Se decidió rellenar la columna "Ciudad" con el valor `"Desconocido"`, ya que se trata de una variable categórica y este reemplazo permite conservar los registros sin eliminar información útil de las unidades. En el caso de "Latitud" y "Longitud", los valores faltantes se sustituyeron por 0, debido a que son variables numéricas y no conviene reemplazarlas con texto, ya que esto podría generar errores en análisis posteriores, visualizaciones o cálculos.

In [15]:
horas.duplicated().sum()
#no tiene duplicados

np.int64(0)

In [16]:
horas.isnull().sum()

Num                0
Distribuidor       0
IMEI               0
Alias              0
Marca              0
Modelo          6263
Ciudad             0
Latitud            0
Longitud           0
2024-01            0
2024-02            0
2024-03            0
2024-04            0
2024-05            0
2024-06            0
2024-07            0
2024-08            0
2024-09            0
2024-10            0
2024-11            0
2024-12            0
dtype: int64

**Ruteo**

In [17]:
ruteo.isnull().sum()

Division                           1
Distancia del Dia                  0
Distancia de Ruta                  0
Cumplimiento Salida (06:30:00)     0
% Utilizacion                      0
% Visitas Plan                     0
% Visitas Clientes                 0
Total Kilos                       35
dtype: int64

In [18]:
ruteo.duplicated().sum()

np.int64(0)

In [19]:
ruteo = ruteo.drop(columns=[
    'Cumplimiento Salida (06:30:00)',
    '% Utilizacion',
    '% Visitas Plan',
    '% Visitas Clientes',
    'Total Kilos'
], errors='ignore')

In [20]:
ruteo.columns

Index(['Division', 'Distancia del Dia', 'Distancia de Ruta'], dtype='str')

En la base de ruteo se revisó la existencia de registros duplicados mediante la función duplicated().sum(). El resultado fue 0, por lo que no se identificaron filas repetidas completas y no fue necesario eliminar duplicados. Posteriormente, se eliminaron las columnas Cumplimiento Salida (06:30:00), % Utilización, % Visitas Plan, % Visitas Clientes y Total Kilos, debido a que no contenían información útil para el análisis, ya que sus valores estaban en cero o como NA. Por esta razón, se conservaron únicamente las variables relevantes para el proyecto: División, Distancia del Día y Distancia de Ruta, las cuales permiten analizar el comportamiento operativo de ruteo por distribuidor.

**Mantenimiento**

In [21]:
mantenimientos.isnull().sum()

No               23904
ALIAS                0
MARCA                4
MODELO               4
CONFIGURACION        4
NO SERIE            38
DISTRIBUIDOR       303
SERVICIO             0
HRMTRO               0
ACTUAL               0
TIPO                 4
FECHA                0
O.T.             14420
ESTATUS              4
Editar           23904
dtype: int64

In [22]:
# Eliminar columnas que no aportan o tienen demasiados nulos
mantenimientos = mantenimientos.drop(columns=[
    'No',
    'Editar',
    'O.T.'
], errors='ignore')

# Rellenar valores nulos en columnas categóricas importantes
columnas_rellenar = [
    'MARCA',
    'MODELO',
    'CONFIGURACION',
    'NO SERIE',
    'DISTRIBUIDOR',
    'TIPO',
    'ESTATUS'
]

for col in columnas_rellenar:
    if col in mantenimientos.columns:
        mantenimientos[col] = mantenimientos[col].fillna('Desconocido')

# Verificar nulos restantes
mantenimientos.isnull().sum()

ALIAS            0
MARCA            0
MODELO           0
CONFIGURACION    0
NO SERIE         0
DISTRIBUIDOR     0
SERVICIO         0
HRMTRO           0
ACTUAL           0
TIPO             0
FECHA            0
ESTATUS          0
dtype: int64

Se eliminaron porque:

No tiene 23,904 nulos y normalmente es solo un consecutivo, entonces no aporta al modelo.

Editar también tiene 23,904 nulos y parece una columna de acción del sistema, no una variable analítica.

O.T. tiene 14,420 nulos. Aunque representa la orden de trabajo, tiene demasiados faltantes y no es tan útil para predecir riesgo o mantenimiento pendiente/fuera.

In [23]:
mantenimientos.duplicated().sum()

np.int64(35)

In [24]:
# Mostrar todas las filas duplicadas completas
duplicados_mantenimientos = mantenimientos[mantenimientos.duplicated(keep=False)]

duplicados_mantenimientos

,ALIAS,MARCA,MODELO,CONFIGURACION,NO SERIE,DISTRIBUIDOR,SERVICIO,HRMTRO,ACTUAL,TIPO,FECHA,ESTATUS
19841,862407063625549,NEW HOLLAND AG,6610S,2WD,S33254M-6099110,MADISA,12900,12860.0,13256,Preventivo,29/08/25 13:59,CerradaFuera
19842,862407063625549,NEW HOLLAND AG,6610S,2WD,S33254M-6099110,MADISA,12900,12860.0,13256,Preventivo,29/08/25 13:59,CerradaFuera
19939,TRACSA - S508004M,NEW HOLLAND AG,7610S HERENCIA,4WD,S508004M-8095782,AGTRAC,3900,3860.0,4266,Preventivo,2025-01-09 13:58:00,Pendiente
19940,TRACSA - S508004M,NEW HOLLAND AG,7610S HERENCIA,4WD,S508004M-8095782,AGTRAC,3900,3860.0,4266,Preventivo,2025-01-09 13:58:00,Pendiente
20370,862407061497149,NEW HOLLAND AG,TS6.120,FWD 8X8 PS,NH11088M-002191675,ATN,1500,1460.0,2111,Preventivo,14/09/25 13:55,Pendiente
...,...,...,...,...,...,...,...,...,...,...,...,...
23657,MADISA-S402871M-8107456,NEW HOLLAND AG,7610S HERENCIA,2WD,S402871M-8107456,MADISA,1800,1760.0,1841,Preventivo,30/12/25 14:00,Pendiente
23804,862407061586479,NEW HOLLAND AG,7610S HERENCIA,LODERO,S509351M-8131368,ATC,1500,1500.0,1762,Preventivo,16/10/25 13:57,CerradaFuera
23805,862407061586479,NEW HOLLAND AG,7610S HERENCIA,LODERO,S509351M-8131368,ATC,1500,1500.0,1762,Preventivo,16/10/25 13:57,CerradaFuera
23894,DASSA-4100041M,NEW HOLLAND AG,TS6.120 ELITE,FWD 8X8 MS CABINA,4100041M-001871869,ARBSA,1500,1460.0,1698,Preventivo,2025-07-11 13:55:00,CerradaFuera


In [25]:
# Eliminar registros duplicados completos en la base de mantenimientos
mantenimientos = mantenimientos.drop_duplicates()

# Verificar que ya no haya duplicados
mantenimientos.duplicated().sum()

np.int64(0)

En la base de mantenimientos se identificaron 35 registros duplicados completos. Al revisar las filas duplicadas, se observó que contenían la misma información en las variables principales del servicio, como ALIAS, MARCA, MODELO, CONFIGURACION, NO SERIE, DISTRIBUIDOR, SERVICIO, HRMTRO, ACTUAL, TIPO, FECHA y ESTATUS. Por esta razón, se decidió eliminar los duplicados mediante drop_duplicates(), ya que mantenerlos podría inflar artificialmente la cantidad de servicios registrados y afectar el análisis de mantenimientos pendientes o fuera de tiempo.

**Population**

In [26]:
population.isnull().sum()

VIN (9 CHARACTERS)                21
VIN (17 CHARACTERS)              312
VIN                                0
BRAND                              0
TYPE                               0
PRODUCT                            0
SERIES                             0
MODEL                              0
TECHNICAL TYPE                     0
ICE CODE                           0
DEVICE SERIAL NUMBER               0
SEVERITY LEVEL                     0
CONTEXT                            0
DEALER                             0
DEALER NAME                      424
DEALER CODE                      284
COUNTRY OF COMPETENCE              0
LATITUDE                           0
LONGITUDE                          0
LAST KNOWN POSITION (COUNTRY)      0
TELEMATICS BOX                     0
TELEMATICS BRAND                   0
TELEMATICS TYPE                    0
TELEMATICS MODEL                   0
OWNER NAME                         0
ENGINE HOURS                       0
LATEST COMMUNICATION DATE          0
B

In [27]:
population = population.drop(columns=[
    'VIN (9 CHARACTERS)',
    'LOWER PIN BREAK',
    'UPPER PIN BREAK',
    'EULA DATE'
], errors='ignore')

Se eliminaron VIN (9 CHARACTERS) porque es una version parcial del VIN. Se guarda VIN 17 porque nos ayudará con severity level después.
LOWER PIN BREAK y UPPER PIN BREAK tienen bastantes nulos y no parecen variables centrales para el análisis de mantenimiento o aftermarket.
EULA DATE tiene muchos nulos y no aporta directamente a predecir mantenimiento pendiente/fuera.

In [28]:
columnas_desconocido = [
    'DEALER NAME',
    'DEALER CODE',
    'MYFLEET DEALER',
    'MYFLEET DEALER NAME',
    'EULA ACCEPTED'
]

for col in columnas_desconocido:
    if col in population.columns:
        population[col] = population[col].fillna('Desconocido')

Por otro lado, se conservaron variables como SEVERITY LEVEL, ENGINE HOURS, DEALER, DEALER NAME, LATITUDE, LONGITUDE, LATEST COMMUNICATION DATE y datos de telemática, ya que estas permiten conectar el estado operativo de la unidad con el riesgo de intervención técnica. Para las columnas categóricas con valores faltantes, como datos de dealer o aceptación de EULA, se reemplazaron los nulos por "Desconocido" con el fin de mantener los registros disponibles sin perder unidades relevantes para el análisis.

In [29]:
population.isnull().sum()

VIN (17 CHARACTERS)              312
VIN                                0
BRAND                              0
TYPE                               0
PRODUCT                            0
SERIES                             0
MODEL                              0
TECHNICAL TYPE                     0
ICE CODE                           0
DEVICE SERIAL NUMBER               0
SEVERITY LEVEL                     0
CONTEXT                            0
DEALER                             0
DEALER NAME                        0
DEALER CODE                        0
COUNTRY OF COMPETENCE              0
LATITUDE                           0
LONGITUDE                          0
LAST KNOWN POSITION (COUNTRY)      0
TELEMATICS BOX                     0
TELEMATICS BRAND                   0
TELEMATICS TYPE                    0
TELEMATICS MODEL                   0
OWNER NAME                         0
ENGINE HOURS                       0
LATEST COMMUNICATION DATE          0
BDE PROCESSED AT TIME              0
M

In [30]:
population.duplicated().sum()

np.int64(50)

In [31]:
duplicados_population = population[population.duplicated(keep=False)]

duplicados_population

,VIN (17 CHARACTERS),VIN,BRAND,TYPE,PRODUCT,SERIES,MODEL,TECHNICAL TYPE,ICE CODE,DEVICE SERIAL NUMBER,...,TELEMATICS BRAND,TELEMATICS TYPE,TELEMATICS MODEL,OWNER NAME,ENGINE HOURS,LATEST COMMUNICATION DATE,BDE PROCESSED AT TIME,MYFLEET DEALER,MYFLEET DEALER NAME,EULA ACCEPTED
29,JAFSV300HRM456279,JAFSV300HRM456279,CASE CONSTRUCTION,LIGHT EQUIPMENT,SKID STEER LOADERS,B Series Skid Steer Loaders,SV300B,550711208,E.S.64.009.014.002,1598-3C000D50,...,CC,SSL,SV300B (T3),POLLO INDUSTRIALIZADO DE MEXICO,1571,2026-02-09 12:03:56,2026-02-09 00:33:33,190-08B288A,Dimapesa Sa De Cv,Yes
30,JAFSV300HRM456279,JAFSV300HRM456279,CASE CONSTRUCTION,LIGHT EQUIPMENT,SKID STEER LOADERS,B Series Skid Steer Loaders,SV300B,550711208,E.S.64.009.014.002,1598-3C000D50,...,CC,SSL,SV300B (T3),POLLO INDUSTRIALIZADO DE MEXICO,1571,2026-02-09 12:03:56,2026-02-09 00:33:33,190-08B288A,Dimapesa Sa De Cv,Yes
41,JJGN58NRLPC787412,JJGN58NRLPC787412,CASE CONSTRUCTION,LIGHT EQUIPMENT,LOADER BACKHOES,N Series,580N,675717050,E.S.65.004.001.050,1598-3C000A36,...,CC,TLB_US,580N (T3),MARIA ELSI CASANOVA ROJAS,3457,2026-02-09 12:59:54,2026-02-09 01:33:33,190-08B288A,Dimapesa Sa De Cv,Yes
42,JJGN58NRLPC787412,JJGN58NRLPC787412,CASE CONSTRUCTION,LIGHT EQUIPMENT,LOADER BACKHOES,N Series,580N,675717050,E.S.65.004.001.050,1598-3C000A36,...,CC,TLB_US,580N (T3),MARIA ELSI CASANOVA ROJAS,3457,2026-02-09 12:59:54,2026-02-09 01:33:33,190-08B288A,Dimapesa Sa De Cv,Yes
43,JJGN58NRTPC787433,JJGN58NRTPC787433,CASE CONSTRUCTION,LIGHT EQUIPMENT,LOADER BACKHOES,N Series,580N,675717050,E.S.65.004.001.050,1598-3C0009C9,...,CC,TLB_US,580N (T3),Johnson Controls,1903,2026-02-08 19:25:28,2026-02-09 00:33:33,190-08B288A,Dimapesa Sa De Cv,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
570,NKJ851NXJSKH46430,NKJ851NXJSKH46430,CASE CONSTRUCTION,LIGHT EQUIPMENT,LOADER BACKHOES,NX Series,851NX,675040251,E.S.65.022.004.003,1603-360029FA,...,CC,TLB_EU,851NX (Stage V),Jehlum Engineers & Construction Equipments,115,2026-02-07 20:38:42,2026-02-09 00:33:33,740-CH8599A,Jehlum Engineers & Construction Equipments,Desconocido
572,PRCY9000ERPA07217,PRCY9000ERPA07217,CASE IH,HARVESTING EQUIPMENT,SUGAR CANE,AUSTOFT 9000 Series,AUSTOFT 9000,579520033,A.J.17.004.001.034,540818555,...,CS,Sugar Cane Harvester,SWSN04,Inter Agrups,12,2026-02-05 12:37:10,2026-02-09 00:33:33,190-08B273A,Agrotractores Del Centro Sa De Cv,Desconocido
573,PRCY9000ERPA07217,PRCY9000ERPA07217,CASE IH,HARVESTING EQUIPMENT,SUGAR CANE,AUSTOFT 9000 Series,AUSTOFT 9000,579520033,A.J.17.004.001.034,540818555,...,CS,Sugar Cane Harvester,SWSN04,Inter Agrups,12,2026-02-05 12:37:10,2026-02-09 00:33:33,190-08B273A,Agrotractores Del Centro Sa De Cv,Desconocido
581,PRCY9900PPPA06591,PRCY9900PPPA06591,CASE IH,HARVESTING EQUIPMENT,SUGAR CANE,AUSTOFT 9000 Series,AUSTOFT 9900,579520036,A.J.17.004.002.030,531248282,...,CS,Sugar Cane Harvester,MY21,ASOCIACION DE CAÑEROS PROPIETARIOS Y EJIDATARIOS,5906,2026-02-09 11:53:09,2026-02-09 00:33:33,190-08B072A,Agro Equipos S.A. De C.V.,Yes


In [32]:
# Eliminar duplicados completos en Population
population = population.drop_duplicates()

# Verificar que ya no haya duplicados
population.duplicated().sum()

np.int64(0)

En la base Population se identificaron registros duplicados completos. Al revisar las filas, se observó que correspondían a unidades repetidas con la misma información en variables clave como VIN, BRAND, TYPE, MODEL, ICE CODE y DEVICE SERIAL NUMBER. Debido a que esta base representa equipos individuales con telemetría, mantener estos duplicados podría alterar el conteo real de unidades, la distribución por severidad y el análisis de horas de motor. Por esta razón, se eliminaron los duplicados completos mediante drop_duplicates().

**Reporte unidades día anterior**

In [33]:
reporte.isnull().sum()

No                         0
IMEI                       0
Alias                      0
Marca                    717
Modelo                   939
Configuracion           1416
Fecha Alta               642
No. Serie                710
Distribuidor            1203
Ciudad                  1203
Estado                  1203
Horometro                  0
Odometro                   0
Total                      0
Cerrados                   0
C.Fuera                    0
Pendientes                 0
50                      4016
300                     4932
600                     5905
900                     6743
1200                    7538
1500                    8196
1800                    8775
2100                    9253
2400                    9644
Cliente                10482
Telefono               10654
Email                  10631
Latitud                  238
Longitud                 238
PEP                    10846
Combustible                0
Distancia                  0
Vel. Maxima (k

In [34]:
# Eliminar columnas que no aportan al análisis final
reporte = reporte.drop(columns=[
    'No',
    'Cliente',
    'Telefono',
    'Email',
    'PEP'
], errors='ignore')

# Rellenar categóricas útiles
columnas_desconocido = [
    'Marca',
    'Modelo',
    'Configuracion',
    'No. Serie',
    'Distribuidor',
    'Ciudad',
    'Estado'
]

for col in columnas_desconocido:
    if col in reporte.columns:
        reporte[col] = reporte[col].fillna('Desconocido')

# Convertir Fecha Alta a formato fecha
reporte['Fecha Alta'] = pd.to_datetime(reporte['Fecha Alta'], errors='coerce')

# Rellenar coordenadas manteniendo formato numérico
reporte['Latitud'] = reporte['Latitud'].fillna(0)
reporte['Longitud'] = reporte['Longitud'].fillna(0)

# Revisar nulos restantes
reporte.isnull().sum()

IMEI                      0
Alias                     0
Marca                     0
Modelo                    0
Configuracion             0
Fecha Alta              642
No. Serie                 0
Distribuidor              0
Ciudad                    0
Estado                    0
Horometro                 0
Odometro                  0
Total                     0
Cerrados                  0
C.Fuera                   0
Pendientes                0
50                     4016
300                    4932
600                    5905
900                    6743
1200                   7538
1500                   8196
1800                   8775
2100                   9253
2400                   9644
Latitud                   0
Longitud                  0
Combustible               0
Distancia                 0
Vel. Maxima (kph)         0
Vel. Promedio (kph)       0
T. Encendido (hrs)        0
T. Apagado (hrs)          0
dtype: int64

En la base reporte, se realizó una limpieza enfocada en conservar las variables más relevantes para el análisis final de riesgo de mantenimiento y oportunidad de aftermarket. Primero, se eliminaron columnas que no aportaban directamente al análisis, como No, Cliente, Telefono, Email y PEP. La columna No funcionaba únicamente como consecutivo, mientras que las columnas de contacto del cliente tenían una alta cantidad de valores faltantes y no eran necesarias para construir indicadores técnicos o predictivos.

Después, se rellenaron con “Desconocido” las variables categóricas importantes, como Marca, Modelo, Configuracion, No. Serie, Distribuidor, Ciudad y Estado. Esto permitió conservar los registros sin perder unidades útiles para el análisis, manteniendo disponible la información descriptiva de cada equipo.

Las columnas Latitud y Longitud se rellenaron con 0 para conservar su formato numérico y evitar errores en visualizaciones o análisis posteriores. Después de la limpieza, la mayoría de las columnas relevantes quedaron sin valores nulos, y únicamente Fecha Alta conserva registros faltantes, lo cual es aceptable porque no se debe inventar una fecha cuando no está registrada.

In [35]:
# Detectar columnas de intervalos aunque estén como texto o número
posibles_intervalos = [50, 300, 600, 900, 1200, 1500, 1800, 2100, 2400]

columnas_intervalos = [
    col for col in reporte.columns
    if col in posibles_intervalos or str(col) in [str(x) for x in posibles_intervalos]
]

# Crear variables derivadas
reporte['intervalos_con_info'] = reporte[columnas_intervalos].notnull().sum(axis=1)
reporte['intervalos_sin_info'] = reporte[columnas_intervalos].isnull().sum(axis=1)

# Revisar que se crearon bien
reporte[['intervalos_con_info', 'intervalos_sin_info']].head()

,intervalos_con_info,intervalos_sin_info
0,0,9
1,0,9
2,0,9
3,0,9
4,0,9


Las columnas correspondientes a los intervalos de mantenimiento preventivo, considerando que estas columnas podían estar guardadas como texto o como números. A partir de ellas, se crearon dos variables derivadas: intervalos_con_info e intervalos_sin_info. La primera indica cuántos intervalos de mantenimiento tienen información registrada por unidad, mientras que la segunda muestra cuántos intervalos permanecen vacíos.

Esta transformación es importante porque permite resumir el estado de seguimiento preventivo de cada unidad en variables más fáciles de interpretar. Aunque los intervalos tengan muchos valores nulos, no se eliminaron porque esos vacíos pueden representar servicios que aún no aplican, unidades que no han alcanzado cierto nivel de horas o falta de registro. Por eso, en lugar de borrar esas columnas, se aprovecharon para construir indicadores que ayudan a medir el nivel de información disponible y el posible rezago de mantenimiento por unidad.

In [36]:
reporte['Fecha Alta'] = pd.to_datetime(reporte['Fecha Alta'], errors='coerce')
reporte['edad_equipo'] = 2026 - reporte['Fecha Alta'].dt.year

In [37]:
reporte[['Fecha Alta', 'edad_equipo']].isnull().sum()

Fecha Alta     642
edad_equipo    642
dtype: int64

Se convirtió la columna Fecha Alta a formato de fecha para poder utilizarla correctamente en el análisis. A partir de esta variable se creó edad_equipo, calculada como la diferencia entre el año de análisis 2026 y el año de alta de cada unidad. Esta nueva variable es importante porque permite analizar si la antigüedad del equipo se relaciona con un mayor riesgo de mantenimiento o con una mayor oportunidad de aftermarket.

In [38]:
reporte.duplicated().sum()
#no tiene duplicados

np.int64(0)

**Manejo de valores atípicos**

Antes de modificar los valores atípicos, se realizó una revisión por base de datos para identificar si existían registros con comportamientos extremos en las variables numéricas. Esta revisión es importante porque, dentro del proyecto CNH, un valor atípico no necesariamente representa un error, también puede indicar una unidad con mayor uso, mayor desgaste, mayor distancia recorrida o mayor necesidad de mantenimiento. Por esta razón, primero se detectaron los posibles atípicos en cada base y posteriormente se evaluó si debían conservarse como información relevante para el análisis final.

In [39]:
def detectar_outliers_iqr(df, nombre_base):
    print(f"\nValores atípicos en la base: {nombre_base}")
    print("-" * 50)

    columnas_numericas = df.select_dtypes(include=['int64', 'float64']).columns

    for col in columnas_numericas:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1

        limite_inferior = Q1 - 1.5 * IQR
        limite_superior = Q3 + 1.5 * IQR

        outliers = df[(df[col] < limite_inferior) | (df[col] > limite_superior)]

        print(f"{col}: {len(outliers)} valores atípicos")

In [40]:
detectar_outliers_iqr(distribuidores, "Distribuidores")



Valores atípicos en la base: Distribuidores
--------------------------------------------------
No.: 0 valores atípicos
Horas: 4 valores atípicos


Se conservan, en la base de distribuidores se identificaron 4 valores atípicos en la variable Horas. Estos registros no se eliminaron, ya que pueden representar distribuidores o unidades con mayor nivel de operación. Al tratarse de pocos casos, se conservaron para no perder información relevante sobre actividad operativa elevada.

In [41]:
detectar_outliers_iqr(horas, "Horas 2024-2025")


Valores atípicos en la base: Horas 2024-2025
--------------------------------------------------
Num: 0 valores atípicos
IMEI: 0 valores atípicos
Latitud: 241 valores atípicos
Longitud: 386 valores atípicos
2024-01: 797 valores atípicos
2024-02: 898 valores atípicos
2024-03: 945 valores atípicos
2024-04: 779 valores atípicos
2024-05: 720 valores atípicos
2024-06: 600 valores atípicos
2024-07: 633 valores atípicos
2024-08: 831 valores atípicos
2024-09: 871 valores atípicos
2024-10: 609 valores atípicos
2024-11: 572 valores atípicos
2024-12: 623 valores atípicos


Se detectaron valores atípicos en las columnas mensuales de actividad. Estos valores no se eliminaron, ya que pueden representar equipos con mayor intensidad de uso en ciertos meses. Dado que el objetivo del proyecto es analizar comportamiento operativo y detectar unidades con posible desgaste o mayor necesidad de mantenimiento, los valores altos de horas se consideran información relevante. Las variables Latitud y Longitud no se interpretaron como atípicos operativos, ya que corresponden a ubicación geográfica y no a desempeño del equipo.

In [42]:
detectar_outliers_iqr(ruteo, "Indicadores de Ruteo")



Valores atípicos en la base: Indicadores de Ruteo
--------------------------------------------------
Distancia del Dia: 2 valores atípicos
Distancia de Ruta: 2 valores atípicos


En la base de ruteo se identificaron 2 valores atípicos en Distancia del Día y 2 en Distancia de Ruta. Estos registros se conservaron porque pueden representar divisiones con mayor recorrido operativo. Al ser variables directamente relacionadas con el uso de la flota, los valores extremos pueden aportar información sobre intensidad de operación.

In [43]:
detectar_outliers_iqr(mantenimientos, "Mantenimientos")



Valores atípicos en la base: Mantenimientos
--------------------------------------------------
SERVICIO: 1038 valores atípicos
HRMTRO: 1051 valores atípicos
ACTUAL: 1077 valores atípicos


En la base de mantenimientos se detectaron valores atípicos en SERVICIO, HRMTRO y ACTUAL. Estos valores no se eliminaron porque pueden representar equipos con mayor acumulación de horas, servicios más avanzados o mayor desgaste operativo.

In [44]:
detectar_outliers_iqr(population, "PopulationView 2026")


Valores atípicos en la base: PopulationView 2026
--------------------------------------------------
TECHNICAL TYPE: 113 valores atípicos
LATITUDE: 95 valores atípicos
LONGITUDE: 63 valores atípicos
ENGINE HOURS: 116 valores atípicos


En PopulationView 2026 se identificaron valores atípicos en variables como ENGINE HOURS, LATITUDE, LONGITUDE y TECHNICAL TYPE. Para el análisis final, se decidió conservar los atípicos de ENGINE HOURS, ya que las horas de motor reflejan intensidad de uso y pueden relacionarse con mayor deterioro operativo. En cambio, LATITUDE y LONGITUDE se consideraron variables de ubicación, por lo que no se interpretaron como atípicos de desempeño. De igual forma, TECHNICAL TYPE no se trató como una variable operativa directa, sino como una característica técnica del equipo.

In [45]:
detectar_outliers_iqr(reporte, "Reporte unidades día anterior")


Valores atípicos en la base: Reporte unidades día anterior
--------------------------------------------------
Horometro: 447 valores atípicos
Odometro: 1368 valores atípicos
Total: 620 valores atípicos
Cerrados: 1536 valores atípicos
C.Fuera: 2234 valores atípicos
Pendientes: 881 valores atípicos
Latitud: 553 valores atípicos
Longitud: 608 valores atípicos
Combustible: 2425 valores atípicos
Distancia: 2374 valores atípicos
Vel. Maxima (kph): 993 valores atípicos
Vel. Promedio (kph): 2197 valores atípicos
T. Encendido (hrs): 2043 valores atípicos
T. Apagado (hrs): 1346 valores atípicos
intervalos_con_info: 0 valores atípicos
intervalos_sin_info: 0 valores atípicos
edad_equipo: 26 valores atípicos


En la base Reporte unidades día anterior se detectaron valores atípicos en variables operativas y de mantenimiento como Horometro, Odometro, Cerrados, C.Fuera, Pendientes, Combustible, Distancia, T. Encendido, T. Apagado y edad_equipo. Estos valores no se eliminaron, ya que pueden representar unidades con alta intensidad de uso, mayor desgaste, mayor número de servicios pendientes o fuera de tiempo, y por lo tanto mayor riesgo de intervención técnica. En lugar de tratarlos como errores, se conservaron como señales relevantes para el análisis predictivo y la identificación de oportunidades de aftermarket.

En el manejo de valores atípicos, se decidió **no eliminar automáticamente los registros extremos detectados por el método IQR**. En este proyecto, los valores atípicos pueden representar unidades con mayor intensidad de uso, mayor desgaste operativo, más servicios pendientes o cerrados fuera, y por lo tanto mayor riesgo de intervención técnica. Por esta razón, los outliers se conservaron como información relevante para el análisis predictivo y comercial.

### Exportar los archivos limpios

In [46]:
distribuidores.to_excel('../data/ArchivosLimpios/new_distribuidor.xlsx', index = False)
horas.to_excel('../data/ArchivosLimpios/new_horas.xlsx', index = False)
ruteo.to_excel('../data/ArchivosLimpios/new_ruteo.xlsx', index = False)
mantenimientos.to_excel('../data/ArchivosLimpios/new_mantenimientos.xlsx', index = False)
population.to_excel('../data/ArchivosLimpios/new_population.xlsx', index = False)
reporte.to_excel('../data/ArchivosLimpios/new_unidades.xlsx', index = False)